### PR TITLE
Let order generators account for selection changes

### LINKED ISSUE #16275: Units group selection issue - Wrong (previously selected) group taking move order
### Issue Summary

As in the title of this topic. Issue appeared in game several OpenRA versions ago, it begins with one update in 2017. I've been waited for solution without rising new issue-bug on GitHub hoping that it's more popular than it's in reality, but non of the newer versions solved the issue.

#### System Information
- **Operating System:** Fedora 28/29; kernel 4.20.13-200.fc29.x86_64 and olders to 4.16.xx-xxx
- **.NET / Mono Version:** 5.18.0.268-0.xamarin.3.epel7.x86_64
- **OpenRA Version:** playtest-20190302 and olders to approx. 2017xxxx
- **Mod:** Red Alert

#### Additional Information:
- Steps to reproduce
	1. Select units and assign to group "1"
	2. Select other units and assign to group "2"
	3. Select "1" group by number key and give attack move order
	4. Select "2" group by number key and give move order - ISSUE (group "1" taking move order instead of selected group "2")

### PR DESCRIPTION
**Pull request by abcdefg30**
Closes #16275.

### CODE REVIEW

**Comment by abcdefg30** [comment]
I think I've tested all the cases now, so this should be ready for review.

File: OpenRA.Mods.Common/Orders/GuardOrderGenerator.cs [hunk 1]
```diff
@@ -28,29 +28,30 @@ protected override IEnumerable<Order> OrderInner(World world, CPos xy, MouseInpu
 				yield break;
 
 			var target = FriendlyGuardableUnits(world, mi).FirstOrDefault();
-			if (target == null || Subjects.All(s => s.IsDead))
```
Comment (pchote) [inline OpenRA.Mods.Common/Orders/GuardOrderGenerator.cs:31]: Are you 100% sure that actors in `Subjects` can't die between `SelectionChanged` being called and this? We haven't ever checked this in `AttackMoveOrderGenerator` so I assume the answer is yes, but it would be good to get confirmation that you have thought about this.

**Review by pchote — COMMENTED** [review]
The idea and implementation here look reasonable, but I haven't yet tested this ingame.

One question, and IMO the two commits really should be merged together (since the second is fixing a bug introduced in the first).

**Comment by pchote** [comment]
Something isn't right with minelayers. Build two, and put them in separate control groups. Select one, start to target a field, switch to the other group, then issue the order. Expected behaviour: new minelayer starts laying the ordered field. Actual behaviour: order is not issued (or is issued but ignored?).

**Comment by pchote** [comment]
The issue is that the minelayer targeting actually uses *two* orders. The `BeginMinefield` order goes to the first minelayer, and the `PlaceMinefield` order to the second. The second minelayer doesn't have an initial value for `minefieldStart` so tries to lay a field from 0,0 to the selected cell. Thanks to #18081, it does nothing. Once that is fixed, it will instead try to lay a bogus field.

This could be fixed by changing minelayers to issue a single order with both the start location (`self.World.Map.CellContaining(order.Target.CenterPosition)`) and end location (`order.ExtraLocation`).

File: OpenRA.Mods.Common/Orders/GuardOrderGenerator.cs (hunk 1 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Orders/GuardOrderGenerator.cs:31]: There was a reason for this change and I think actors can't die in-between, but don't remember anymore.

**Comment by abcdefg30** [comment]
Updated. I went with the simple solution of sending the start location with the second order. Merging the two orders had the nasty problems of being a lot of changes and having no good way to distinguish start and end orders any more. (That is because `ExtraLocation` is always set and hence we would have to assume that an end location of (0,0) means this is a minefield start order... The proper fix here would be giving the `Order` class something like a `HasExtraLocation` property.)

**Comment by teinarss** [comment]
One inconsistentcy that i found was that all orders keep the ordergenerator except `PortableChrono`, which cancel it.

**Comment by abcdefg30** [comment]
This is because of how PortableChrono works: Only with a single unit ever.

**Review by teinarss — DISMISSED** [review]

File: OpenRA.Mods.Common/Orders/GuardOrderGenerator.cs [hunk 2]
```diff
@@ -28,29 +28,30 @@ protected override IEnumerable<Order> OrderInner(World world, CPos xy, MouseInpu
...
-			yield return new Order(OrderName, null, Target.FromActor(target), queued, null, Subjects.Where(s => s != target).ToArray());
+			yield return new Order(OrderName, null, Target.FromActor(target), queued, null, subjects.Where(s => s != target).ToArray());
 		}
 
-		public override void Tick(World world)
+		public override void SelectionChanged(World world, IEnumerable<Actor> selected)
 		{
-			if (Subjects.All(s => s.IsDead || !s.Info.HasTraitInfo<GuardInfo>()))
+			subjects = selected.Where(s => s.Info.HasTraitInfo<GuardInfo>());
+			if (!subjects.Any(s => s.Info.HasTraitInfo<AutoTargetInfo>()))
```
Comment (pchote) [inline OpenRA.Mods.Common/Orders/GuardOrderGenerator.cs:44]: Assigning subjects based on `GuardInfo` and then checking `AutoTargetInfo` here looks like a typo. If this is intentional, can we have a comment explaining that?

File: OpenRA.Mods.Common/Traits/AttackMove.cs [hunk 3]
```diff
@@ -121,9 +122,13 @@ protected virtual IEnumerable<Order> OrderInner(World world, CPos cell, MouseInp
...
 
-		public override void Tick(World world)
+		public override void SelectionChanged(World world, IEnumerable<Actor> selected)
 		{
-			if (subjects.All(s => s.Actor.IsDead))
+			subjects = selected.SelectMany(a => a.TraitsImplementing<AttackMove>()
+					.Select(am => new TraitPair<AttackMove>(a, am)))
+				.ToArray();
+
+			if (!subjects.Any(s => s.Actor.Info.HasTraitInfo<AutoTargetInfo>()))
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/AttackMove.cs:131]: Assigning subjects based on `AttackMove` and then checking `AutoTargetInfo` here looks like a typo. If this is intentional, can we have a comment explaining that?

File: OpenRA.Mods.Common/Orders/GuardOrderGenerator.cs (hunk 2 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Orders/GuardOrderGenerator.cs:44]: `Guard` uses `AttackMove` which doesn't work without `AutoTarget`. It's written like this because of
```
[12:01:43] | (abcdefg30)  we have Guard and AttackMove on a bunch of actors that can't activate  those because the command bar requires them to have AutoTarget as well
[12:02:01] | (abcdefg30) but you can activate them when you select them together with another actor that has the trait + AutoTarget
[12:02:16] | (pchote) iirc the idea was so that they would follow along if in a group with others
```

For which I later filed #18004 and #18005.

Tl;dr: This is to keep current behaviour.

File: OpenRA.Mods.Common/Traits/AttackMove.cs (hunk 3 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/AttackMove.cs:131]: See above.

**Comment by abcdefg30** [comment]
Added comments.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/df3c8030d0545e620639710959e08b7f180882ba
